OSSM-8507 Migration guide for multitenant 2.6 (#86917)
openshift/openshift-docs: modules/ossm-migrating-a-multitenant-deployment.adoc

// Module included in the following assemblies:
//
// * service-mesh-docs-main/about/ossm-migrating-migration-guides-assembly.adoc

:_mod-docs-content-type: PROCEDURE
[id="migrating-a-multitenant-deployment_{context}""]
== Migrating a multitenant deployment

.Prerequisites

> * You have deployed a cluster on {ocp-product-title} 4.14 or later.

Reviewer: ```suggestion
* You have deployed {ocp-product-title} 4.14 or later.
```

* You are logged in to the {ocp-product-title} web console as a user with the cluster-admin role.
* You have completed the premigration checklists.
* You have the {SMProduct} {SMv2Version} Operator installed.
* You have the {SMProduct} 3 Operator installed.
* You have `IstioCNI` installed.
* You have `istioctl` installed.
* You are running a `MultiTenant` `ServiceMeshControlPlane`.

> 

Reviewer 2: You have the 'bookinfo' app installed. 

Please, can you add this line

.Procedure

The `bookinfo` example application is being used for demonstration purposes. You can follow these same steps with your own workloads.

> . Create your `Istio` resource.

Reviewer: Can we include the note as it's done in https://github.com/openshift-service-mesh/sail-operator/tree/main/docs/ossm/ossm2-migration/multi-tenancy#install-openshift-service-mesh-30 ?
NOTE: For simplicity, we are using a minimal example for Istio resource. Read [SMCP to Istio mapping](https://github.com/openshift-service-mesh/sail-operator/tree/main/docs/ossm/ossm2-migration/multi-tenancy#TODO) to see how to transform fields used in SMCP resource to fields in Istio resource.

Author: Phrasing about minimal example has been added, along with an "Additional resources" section that will include a link to the field mappings when that PR has been reviewed and merged.

+
.Example `Istio` resource
+
[source,yaml]
----
   apiVersion: sailoperator.io/v1alpha1
   kind: Istio
   metadata:
     name: istio-tenant-a
   spec:
     namespace: istio-system-tenant-a <1>

>      version: v1.23.0

Reviewer: Is it possible to use some variable? Istio version for GA will be probably 1.24.3

>    apiVersion: sailoperator.io/v1alpha1

Reviewer: Ah yes, that will probably deserve dedicated ticket. For GA all examples using `sailoperator.io/v1alpha1` will have to be updated to use `sailoperator.io/v1`

>    apiVersion: sailoperator.io/v1alpha1

Author: There's  Jira for updating this across OSSM 3 prod docs: [OSSM-8754](https://issues.redhat.com/browse/OSSM-8754)

Thanks for reminding me!

   #       - name: prometheus
   #         prometheus: {}
   #       - name: otel
   #         opentelemetry:
   #           port: 4317
   #           service: otel-collector.opentelemetrycollector-3.svc.cluster.local
----
+
<1> The `spec.namespace` field in your `Istio` resource must be the **same** namespace as your `ServiceMeshControlPlane`. If you set the `spec.namespace` field in your `Istio` resource to a different namespace than your `ServiceMeshControlPlane`, the migration will not work properly.
<2> controlplanes by default watch the entire cluster and when managing multiple controlplanes on a single cluster, you must narrow the scope of each controlplane by setting `discoverySelectors`. In this example, we use the label `tenant` but you can use any label or combination of labels that you choose.
<3> If you are migrating metrics and tracing, uncomment and update the `extensionProviders` fields according to your tracing and metrics configurations.

. Add your `tenant` label to each one of your dataplane namespaces by running the following command for each dataplane namespace:
+
[source,terminal]
----
$ oc label ns bookinfo tenant=tenant-a
----
+
[NOTE]
====

> With {SMProduct} 2.6, namespaces were enrolled into the mesh by adding them to the `Service MeshMemberRoll` resource. In {SMProduct} 3, you must label each one of your dataplane namespaces to match your `discoverySelectors`.

Reviewer 3: ```suggestion
With {SMProduct} 2.6, namespaces were enrolled into the mesh by adding them to the `ServiceMeshMemberRoll` resource. In {SMProduct} 3, you must label each one of your dataplane namespaces to match your `discoverySelectors`.
```

====

=== Migrating workloads

Now you can migrate your workloads from the {SMProduct} 2.6 controlplane to the {SMproduct} 3.0 controlplane.

[NOTE]
====

> You can migrate workloads and gateways separately.

Reviewer 3: Maybe also add "and in any order" so it's more clear that you don't need to upgrade workloads _before_ you upgrade gateways.
```suggestion
You can migrate workloads and gateways separately.
```

   reviews-v1-686cd989df-9x59z.bookinfo              Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED       NOT SENT     istiod-basic-6c9f8d9894-sh6lx     1.20.8
   reviews-v2-785b8b48fc-l7xkj.bookinfo              Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED       NOT SENT     istiod-basic-6c9f8d9894-sh6lx     1.20.8
   reviews-v3-67889ffd49-7bhxn.bookinfo              Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED       NOT SENT     istiod-basic-6c9f8d9894-sh6lx     1.20.8
----

.. Check which proxies have been migrated to the new {SMProduct} 3.0 controlplane by running the following command:
+
[source,terminal]
----
istioctl ps --istioNamespace istio-system-tenant-a --revision istio-tenant-a
----
+
.Example output
+
[source, terminal]
----
NAME                                         CLUSTER        CDS        LDS        EDS        RDS        ECDS     ISTIOD                      VERSION
productpage-v1-7745c5cc94-wpvth.bookinfo     Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED              istiod-5bbf98dccf-n8566     1.24.3
----

> . Verify the `bookinfo` application is still working correctly by running the following command:

Reviewer 3: Trying to make it more obvious to readers that they will need to verify their own applications are working correctly and the exact commands they will run to do that will be different. The following command only applies to the bookinfo example.
```suggestion
. Verify your application is still working correctly. For the `bookinfo` application, run the following command:
```
